reproject/interpolation/core.py

```diff
@@ -40,28 +40,109 @@ def map_coordinates(image, coords, **kwargs):
+
+    # Check whether celestial components are present
+    if wcs_in.has_celestial and wcs_out.has_celestial:
+        has_celestial = True
+    elif wcs_in.has_celestial:
+        raise ValueError("Input WCS has celestial components but output WCS does not")
+    elif wcs_out.has_celestial:
+        raise ValueError("Output WCS has celestial components but input WCS does not")
+    else:
+        has_celestial = False
+
+    # Check whether a spectral component is present, and if so, check that
+    # the CTYPEs match.
+    if wcs_in.wcs.spec >= 0 and wcs_out.wcs.spec >= 0:
+        if wcs_in.wcs.ctype[wcs_in.wcs.spec] != wcs_out.wcs.ctype[wcs_out.wcs.spec]:
+            raise ValueError("The input ({0}) and output ({1}) spectral "
+                             "coordinate types are not equivalent."
+                             .format(wcs_in.wcs.ctype[wcs_in.wcs.spec],
+                                     wcs_out.wcs.ctype[wcs_out.wcs.spec]))
+    elif wcs_in.wcs.spec >= 0:
+        raise ValueError("Input WCS has a spectral component but output WCS does not")
+    elif wcs_out.wcs.spec >= 0:
+        raise ValueError("Output WCS has a spectral component but input WCS does not")
+
+    # We need to make sure that either the axis types match exactly, or that
+    # they are shuffled but otherwise they are unique and there is a one-to-one
+    # mapping from the input to the output WCS.
+    if tuple(wcs_in.wcs.axis_types) == tuple(wcs_out.wcs.axis_types):
+        needs_reorder = False
+    else:
+        if sorted(wcs_in.wcs.axis_types) == sorted(wcs_in.wcs.axis_types):
+            if len(set(wcs_in.wcs.axis_types)) < wcs_in.wcs.naxis or \
+               len(set(wcs_out.wcs.axis_types)) < wcs_out.wcs.naxis:
+                raise ValueError("axis_types contains non-unique elements, and "
+                                 "input order does not match output order")
+            else:
+                needs_reorder = True
+        else:
+            raise ValueError("axis_types do not map from input WCS to output WCS")
+
+    # Determine mapping from output to input WCS
+    if needs_reorder:
+        axis_types_in = tuple(wcs_out.wcs.axis_types)
+        axis_types_out = tuple(wcs_out.wcs.axis_types)
+        indices_out = [axis_types_out.index(axis_type) for axis_type in axis_types_in]
+    else:
+        indices_out = list(range(wcs_out.wcs.naxis))
 
-    p_out = np.meshgrid(*p_out_ax, indexing='ij')
+    # Check that the units match
+    for index_in, index_out in enumerate(indices_out):
+        unit_in = wcs_in.wcs.cunit[index_in]
+        unit_out = wcs_out.wcs.cunit[index_out]
+        print(unit_in, unit_out)
+        if unit_in != unit_out:
+            raise ValueError("Units differ between input ({0}) and output "
+                             "({1}) WCS".format(unit_in, unit_out))
+
+    # Generate pixel coordinates of output image. This is reversed because
+    # numpy and wcs index in opposite directions.
```

Comment:
maybe add the note that you're reversing [z,y,x] -> [x,y,z], not reversing the actual indices